mlir/lib/Quantum/Transforms/DisentangleSWAP.cpp

```diff
@@ -0,0 +1,216 @@
+        builder.setInsertionPointAfter(insert_after_gate);
+        quantum::CustomOp newGate =
+            builder.create<quantum::CustomOp>(loc,
+                                              /*out_qubits=*/mlir::TypeRange({outQubit.getType()}),
+                                              /*out_ctrl_qubits=*/mlir::TypeRange(),
+                                              /*params=*/mlir::ValueRange(),
+                                              /*in_qubits=*/mlir::ValueRange({inQubit}),
+                                              /*gate_name=*/gateName,
+                                              /*adjoint=*/nullptr,
+                                              /*in_ctrl_qubits=*/mlir::ValueRange(),
+                                              /*in_ctrl_values=*/mlir::ValueRange());
+    
+        return newGate;
+    }
+
+
+    // quantum::CustomOp createSimpleTwoBitGate(StringRef gateName, 
+    //                                          const Value &controlIn, const Value &TargetIn, 
+    //                                          const Value &controlOut, const Value &targetOut, 
+    //                                          mlir::IRRewriter &builder, Location &loc, 
+    //                                          const quantum::CustomOp &insert_after_gate)
+    // {
+    //     OpBuilder::InsertionGuard insertionGuard(builder);
+    //     builder.setInsertionPointAfter(insert_after_gate);
+    //     quantum::CustomOp newGate =
+    //         builder.create<quantum::CustomOp>(loc,
+    //                                           /*out_qubits=*/mlir::TypeRange({targetOut.getType()}),
+    //                                           /*out_ctrl_qubits=*/mlir::TypeRange({controlOut.getType()}),
+    //                                           /*params=*/mlir::ValueRange(),
+    //                                           /*in_qubits=*/mlir::ValueRange({TargetIn}),
+    //                                           /*gate_name=*/gateName,
+    //                                           /*adjoint=*/nullptr,
+    //                                           /*in_ctrl_qubits=*/mlir::ValueRange({controlIn}),
+    //                                           /*in_ctrl_values=*/mlir::ValueRange());
+
+    //     return newGate;
+    // }
+
+    bool canScheduleOn(RegisteredOperationName opInfo) const override
+    {
+        return opInfo.hasInterface<FunctionOpInterface>();
+    }
+
+    void runOnOperation() override
+    {
+        LLVM_DEBUG(dbgs() << "disentangle SWAP pass\n");
+
+        func::FuncOp func = cast<func::FuncOp>(getOperation());
+        mlir::IRRewriter builder(func->getContext());
+        Location loc = func->getLoc();
+
+        if (func.getSymName() != FuncNameOpt) {
+            // not the function to run the pass on
+            return;
+        }
+
+        ///////////////////////////
+
+        PropagateSimpleStatesAnalysis &pssa = getAnalysis<PropagateSimpleStatesAnalysis>();
```

Comment:
We have a [tutorial](https://docs.pennylane.ai/projects/catalyst/en/latest/dev/transforms.html) for the basic catalyst mlir structure, and how to write a mlir pass. The tutorial covers more than what we need to know here. Here we just need to know two things:

1. The `canScheduleOn` method above determines that the pass will only be applied to function operations, i.e. things that look like `func.func ... {...}` in the IR
2. The `runOnOperation` method here determines what transformations will be performed on the target operations.

In other words, for each function, the `runOnOperation` method happens once.

Let's take a look at the [construction of the `PropagateSimpleStatesAnalysis` object](https://github.com/PennyLaneAI/catalyst/blob/qstate_mlir_interpreter_fsm/mlir/lib/Quantum/Transforms/PropagateSimpleStatesAnalysis.hpp). The object carries a map, which records the abstract enum qubit state of each qubit value. The write into the map happens upon the analysis object's construction. If the analysis object is only constructed once, the map will be written into only once. 

Say you perform some transformation that adds a new gate. At the time of the analysis object's creation above (when the pass is run on the function, but before the walk visitor has created the new gate op), the new gate wouldn't have existed yet, so it wouldn't be in the map. 

However, looking at your code, it seems like you haven't created any gates yet. It would be helpful to see your test case! Note that the existing `PropagateSimpleStatesAnalysis` is only my first draft so there could be bugs : )

What might be helpful is, since the assertion failure is saying some qubit values don't have corresponding states in the map, before the assertion line, we can print out the operation being analyzed, and print out the map, to see what's going on. In mlir you can print via the following:
```
Thing thing_I_want_to_print;
llvm::errs() << thing_I_want_to_print;
```

Don't hesitate to let me know if you get any progress!